File: Packages/HeistEngine/Sources/HeistCore/Navigation/TrajectoryBuilder.swift
import Foundation

/// Converts a legal path polyline into a denser, rounded locomotion trajectory.
/// Every inserted segment is checked against the same walkability grid, so
/// visual smoothness can never cut a corner the character body cannot clear.
public enum TrajectoryBuilder {
    /// Rounds authored/pathfinder corners and, when there is enough legal
    /// floor, joins the replacement route to the actor's current heading with
    /// a cubic curve. This is the navigation equivalent of steering into a
    /// bend: a replan does not first rotate the body to the new first link.
    ///
    /// Near reversals deliberately remain pivots. A forward-moving curve for a
    /// 180-degree change would need an arbitrary side and a large amount of
    /// floor, so the explicit turnaround locomotion block is more truthful.
    public static func continuous(
        start: WorldPoint,
        path: PathResult,
        initialFacing: Double,
        in grid: NavGrid,
        character: CharacterProfile
    ) -> PathResult {
        let roundedPath = rounded(start: start, path: path, in: grid, character: character)
        let points = [start] + roundedPath.waypoints
        guard points.count > 1 else { return roundedPath }

        let routeFacing = PatrolRoute.yaw(from: points[0], to: points[1])
        let headingChange = abs(PatrolRoute.shortestTurn(from: initialFacing, to: routeFacing))
        guard headingChange >= 4, headingChange <= 110 else { return roundedPath }

        let totalLength = polylineLength(points)
        guard totalLength > max(grid.cellSize * 4, 0.35) else { return roundedPath }

        // Roughly two thirds of one second at the authored walking pace, but
        // never consume most of a short route. The character's comfort radius
        // keeps this scale consistent when profiles change.
        let preferredBlend = max(
            character.preferredCornerRadius * 2.4,
            character.walkSpeed * 0.65
        )
        let maximumBlend = min(preferredBlend, totalLength * 0.55)
        let minimumBlend = max(grid.cellSize * 4, character.preferredCornerRadius)
        guard maximumBlend >= minimumBlend else { return roundedPath }

        let radians = initialFacing * .pi / 180
        let startDirection = WorldPoint(x: sin(radians), y: 0, z: cos(radians))

        // A curve may bulge into a nearby wall. Try progressively shorter
        // joins, accepting one only after every chord passes the same eroded
        // navigation grid as the original route.
        for scale in [1.0, 0.82, 0.66, 0.5] {
            let blendDistance = maximumBlend * scale
            guard blendDistance >= minimumBlend,
                  let join = split(points: points, atDistance: blendDistance) else { continue }

            let joinDirection = (join.segmentEnd - join.segmentStart).normalized
            let handle = blendDistance * 0.38
            let control1 = start + startDirection * handle
            let control2 = join.point - joinDirection * handle
            let sampleCount = max(
                16,
                Int((blendDistance / max(grid.cellSize * 0.3, 0.025)).rounded(.up))
            )
            var curve: [WorldPoint] = []
            for sample in 1...sampleCount {
                let t = Double(sample) / Double(sampleCount)
                let inverse = 1 - t
                curve.append((
                    start * (inverse * inverse * inverse)
                        + control1 * (3 * inverse * inverse * t)
                        + control2 * (3 * inverse * t * t)
                        + join.point * (t * t * t)
                ).onFloorPlane)
            }

            var previous = start
            let legal = curve.allSatisfy { point in
                defer { previous = point }
                return PathFinder.hasLineOfSight(from: previous, to: point, in: grid)
            }
            guard legal else { continue }

            let stitched = [start] + curve + join.remaining.dropFirst()
            return result(for: stitched, excludingStart: true)
        }
        return roundedPath
    }

    public static func rounded(
        start: WorldPoint,
        path: PathResult,
        in grid: NavGrid,
        character: CharacterProfile
    ) -> PathResult {
        var source = [start] + path.waypoints
        source = source.reduce(into: []) { result, point in
            if result.last?.planarDistance(to: point) ?? .greatestFiniteMagnitude > 1e-6 {
                result.append(point.onFloorPlane)
            }
        }
        guard source.count > 2, character.preferredCornerRadius > 0 else {
            return result(for: source, excludingStart: true)
        }

        var output = [source[0]]
        for index in 1..<(source.count - 1) {
            let a = source[index - 1]
            let corner = source[index]
            let c = source[index + 1]
            let incoming = a - corner
            let outgoing = c - corner
            let incomingLength = incoming.planarLength
            let outgoingLength = outgoing.planarLength
            guard incomingLength > 1e-6, outgoingLength > 1e-6 else { continue }

            let dot = max(-1, min(1,
                incoming.normalized.x * outgoing.normalized.x
                    + incoming.normalized.z * outgoing.normalized.z
            ))
            // Direction vectors both point away from the corner. -1 is a
            // straight continuation and needs no arc.
            if dot < -0.995 {
                output.append(corner)
                continue
            }

            var radius = min(
                character.preferredCornerRadius,
                incomingLength * 0.42,
                outgoingLength * 0.42
            )
            var accepted: [WorldPoint]?
            for _ in 0..<5 where radius >= grid.cellSize {
                let entry = corner + incoming.normalized * radius
                let exit = corner + outgoing.normalized * radius
                let samples = max(4, Int((radius * 2 / max(grid.cellSize, 0.01)).rounded(.up)))
                var candidate: [WorldPoint] = [entry]
                for sample in 1...samples {
                    let t = Double(sample) / Double(samples)
                    let inverse = 1 - t
                    candidate.append(
                        (entry * (inverse * inverse)
                            + corner * (2 * inverse * t)
                            + exit * (t * t)).onFloorPlane
                    )
                }

                var previous = output.last ?? a
                let legal = candidate.allSatisfy { point in
                    defer { previous = point }
                    return PathFinder.hasLineOfSight(from: previous, to: point, in: grid)
                }
                if legal {
                    accepted = candidate
                    break
                }
                radius *= 0.5
            }

            if let accepted {
                output.append(contentsOf: accepted)
            } else {
                output.append(corner)
            }
        }
        output.append(source[source.count - 1])
        return result(for: output, excludingStart: true)
    }

    private static func result(for points: [WorldPoint], excludingStart: Bool) -> PathResult {
        let waypoints = excludingStart ? Array(points.dropFirst()) : points
        let length = zip(points, points.dropFirst()).reduce(0) {
            $0 + $1.0.planarDistance(to: $1.1)
        }
        return PathResult(waypoints: waypoints, length: length)
    }

    private struct PolylineSplit {
        var point: WorldPoint
        var segmentStart: WorldPoint
        var segmentEnd: WorldPoint
        var remaining: [WorldPoint]
    }

    private static func split(points: [WorldPoint], atDistance requested: Double) -> PolylineSplit? {
        var remainingDistance = max(0, requested)
        for index in 0..<(points.count - 1) {
            let from = points[index]
            let to = points[index + 1]
            let length = from.planarDistance(to: to)
            guard length > 1e-9 else { continue }
            if remainingDistance <= length {
                let point = (from + (to - from) * (remainingDistance / length)).onFloorPlane
                return PolylineSplit(
                    point: point,
                    segmentStart: from,
                    segmentEnd: to,
                    remaining: [point] + Array(points.dropFirst(index + 1))
                )
            }
            remainingDistance -= length
        }
        return nil
    }

    private static func polylineLength(_ points: [WorldPoint]) -> Double {
        zip(points, points.dropFirst()).reduce(0) {
            $0 + $1.0.planarDistance(to: $1.1)
        }
    }
}

File: UnityPort/Assets/DerClou/Core/Navigation/TrajectoryBuilder.cs
namespace DerClou.Core.Navigation
{
    using System;
    using System.Collections.Generic;
    using DerClou.Core.Data;

    /// Deterministic port of the validated Swift TrajectoryBuilder. It rounds
    /// legal corners and joins an ordinary retarget to the actor's current
    /// heading with a cubic curve. Every inserted chord is checked against
    /// the same authoritative grid, so smoothness can never cut a wall.
    public static class TrajectoryBuilder
    {
        public static WorldPoint[] Continuous(
            WorldPoint start,
            WorldPoint[] path,
            float initialFacing,
            NavGrid grid,
            CharacterProfile character)
        {
            var rounded = Rounded(start, path, grid, character);
            var points = WithStart(start, rounded);
            if (points.Count <= 1) return rounded;

            float routeFacing = Yaw(points[0], points[1]);
            float headingChange = MathF.Abs(DeltaAngle(initialFacing, routeFacing));
            if (headingChange < 4f || headingChange > 110f) return rounded;

            float totalLength = PolylineLength(points);
            if (totalLength <= MathF.Max(grid.cellSize * 4f, 0.35f)) return rounded;

            float preferredBlend = MathF.Max(character.preferredCornerRadius * 2.4f, character.walkSpeed * 0.65f);
            float maximumBlend = MathF.Min(preferredBlend, totalLength * 0.55f);
            float minimumBlend = MathF.Max(grid.cellSize * 4f, character.preferredCornerRadius);
            if (maximumBlend < minimumBlend) return rounded;

            float radians = initialFacing * MathF.PI / 180f;
            var startDirection = new WorldPoint(MathF.Sin(radians), 0, MathF.Cos(radians));
            float[] scales = { 1f, 0.82f, 0.66f, 0.5f };
            foreach (float scale in scales)
            {
                float blendDistance = maximumBlend * scale;
                if (blendDistance < minimumBlend || !TrySplit(points, blendDistance, out var split)) continue;
                var joinDirection = Normalized(Subtract(split.SegmentEnd, split.SegmentStart));
                float handle = blendDistance * 0.38f;
                var control1 = Add(start, Multiply(startDirection, handle));
                var control2 = Subtract(split.Point, Multiply(joinDirection, handle));
                int sampleCount = Math.Max(16, (int)MathF.Ceiling(blendDistance / MathF.Max(grid.cellSize * 0.3f, 0.025f)));
                var curve = new List<WorldPoint>();
                var previous = start;
                bool legal = true;
                for (int sample = 1; sample <= sampleCount; sample++)
                {
                    float t = (float)sample / sampleCount, inv = 1f - t;
                    var point = Add(
                        Add(Multiply(start, inv * inv * inv), Multiply(control1, 3f * inv * inv * t)),
                        Add(Multiply(control2, 3f * inv * t * t), Multiply(split.Point, t * t * t)));
                    point.y = 0;
                    if (!PathFinder.HasLineOfSight(grid, previous, point)) { legal = false; break; }
                    curve.Add(point);
                    previous = point;
                }
                if (!legal) continue;
                curve.AddRange(split.Remaining.GetRange(1, split.Remaining.Count - 1));
                return curve.ToArray();
            }
            return rounded;
        }

        public static WorldPoint[] Rounded(WorldPoint start, WorldPoint[] path, NavGrid grid, CharacterProfile character)
        {
            var source = WithStart(start, path);
            if (source.Count <= 2 || character.preferredCornerRadius <= 0f) return WithoutStart(source);
            var output = new List<WorldPoint> { source[0] };
            for (int i = 1; i < source.Count - 1; i++)
            {
                var a = source[i - 1]; var corner = source[i]; var c = source[i + 1];
                var incoming = Subtract(a, corner); var outgoing = Subtract(c, corner);
                float inLength = Length(incoming), outLength = Length(outgoing);
                if (inLength <= 0.000001f || outLength <= 0.000001f) continue;
                incoming = Normalized(incoming); outgoing = Normalized(outgoing);
                float dot = Clamp(incoming.x * outgoing.x + incoming.z * outgoing.z, -1f, 1f);
                if (dot < -0.995f) { output.Add(corner); continue; }

                float radius = MathF.Min(character.preferredCornerRadius, MathF.Min(inLength * 0.42f, outLength * 0.42f));
                List<WorldPoint> accepted = null;
                for (int attempt = 0; attempt < 5 && radius >= grid.cellSize; attempt++, radius *= 0.5f)
                {
                    var entry = Add(corner, Multiply(incoming, radius));
                    var exit = Add(corner, Multiply(outgoing, radius));
                    int samples = Math.Max(4, (int)MathF.Ceiling(radius * 2f / MathF.Max(grid.cellSize, 0.01f)));
                    var candidate = new List<WorldPoint> { entry };
                    for (int sample = 1; sample <= samples; sample++)
                    {
                        float t = (float)sample / samples, inv = 1f - t;
                        var point = Add(Add(Multiply(entry, inv * inv), Multiply(corner, 2f * inv * t)), Multiply(exit, t * t));
                        point.y = 0; candidate.Add(point);
                    }
                    var previous = output[output.Count - 1]; bool legal = true;
                    foreach (var point in candidate)
                    {
                        if (!PathFinder.HasLineOfSight(grid, previous, point)) { legal = false; break; }
                        previous = point;
                    }
                    if (legal) accepted = candidate;
                    if (accepted != null) break;
                }
                if (accepted != null) output.AddRange(accepted); else output.Add(corner);
            }
            output.Add(source[source.Count - 1]);
            return WithoutStart(output);
        }

        private struct Split { public WorldPoint Point, SegmentStart, SegmentEnd; public List<WorldPoint> Remaining; }
        private static bool TrySplit(List<WorldPoint> points, float requested, out Split result)
        {
            float remaining = MathF.Max(0, requested);
            for (int i = 0; i < points.Count - 1; i++)
            {
                float length = Distance(points[i], points[i + 1]);
                if (length <= 0.000001f) continue;
                if (remaining <= length)
                {
                    var point = Add(points[i], Multiply(Subtract(points[i + 1], points[i]), remaining / length));
                    var tail = new List<WorldPoint> { point };
                    for (int j = i + 1; j < points.Count; j++) tail.Add(points[j]);
                    result = new Split { Point = point, SegmentStart = points[i], SegmentEnd = points[i + 1], Remaining = tail };
                    return true;
                }
                remaining -= length;
            }
            result = default; return false;
        }

        private static List<WorldPoint> WithStart(WorldPoint start, WorldPoint[] path)
        {
            var result = new List<WorldPoint> { start };
            if (path != null) foreach (var p in path) if (Distance(result[result.Count - 1], p) > 0.000001f) result.Add(p);
            return result;
        }
        private static WorldPoint[] WithoutStart(List<WorldPoint> points) { if (points.Count <= 1) return Array.Empty<WorldPoint>(); points.RemoveAt(0); return points.ToArray(); }
        private static float PolylineLength(List<WorldPoint> p) { float v = 0; for (int i = 1; i < p.Count; i++) v += Distance(p[i - 1], p[i]); return v; }
        private static float Yaw(WorldPoint a, WorldPoint b) => MathF.Atan2(b.x - a.x, b.z - a.z) * 180f / MathF.PI;
        public static float DeltaAngle(float current, float target) { float d = (target - current) % 360f; if (d > 180f) d -= 360f; if (d < -180f) d += 360f; return d; }
        private static float Distance(WorldPoint a, WorldPoint b) => Length(Subtract(a, b));
        private static float Length(WorldPoint p) => MathF.Sqrt(p.x * p.x + p.z * p.z);
        private static WorldPoint Normalized(WorldPoint p) { float l = Length(p); return l > 0.000001f ? Multiply(p, 1f / l) : default; }
        private static WorldPoint Add(WorldPoint a, WorldPoint b) => new WorldPoint(a.x + b.x, a.y + b.y, a.z + b.z);
        private static WorldPoint Subtract(WorldPoint a, WorldPoint b) => new WorldPoint(a.x - b.x, a.y - b.y, a.z - b.z);
        private static WorldPoint Multiply(WorldPoint p, float s) => new WorldPoint(p.x * s, p.y * s, p.z * s);
        private static float Clamp(float v, float lo, float hi) => MathF.Max(lo, MathF.Min(hi, v));
    }
}
